[Observability Onboarding] Embed existing logs flows

thomheymann opened the pull request:
Resolves #178948

## Summary

Embeds existing custom and system log flows into new onboarding page. 

Note: As discussed there are several usability issues with this approach which would need to be addressed before this can be merged into main. This PR acts as more of a starting point rather than a finished feature. 

## Screenshots

<img width="1372" alt="Screenshot 2024-04-04 at 17 10 46" src="https://github.com/elastic/kibana/assets/190132/21281d93-ee61-437e-9ef5-249c49c29e75">

## Testing

The experimental onboarding flow can now be accessed using the hidden URL: `/app/experimental-onboarding`

thomheymann committed:
[Observability Onboarding] Link all “Add data” buttons in observability to onboarding landing page (#179459)

Resolves [#178944](https://github.com/elastic/kibana/issues/178944)

## :notebook: Summary

All "Add data" links within Observability link to
`/app/observabilityOnboarding` when on serverless observability.

## :heavy_check_mark: Acceptance criteria
All "Add data" links within Observability link to
`/app/observabilityOnboarding` when on serverless observability:
- Infrastructure > Inventory
- Infrastructure > Hosts

flash1293 committed:
Merge branch 'main' into feat/obs-onboarding-flow

Commit:
Embed system and custom log flows

Commit:
Merge branch 'feat/obs-onboarding-flow' of https://github.com/elastic/kibana into 178948-embed-logs-flow

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

flash1293 commented:
Thanks for this, @thomheymann 

Some comments:
* Honestly even the current state doesn't look too bad to me - the scrollable content in the modal is somewhat mitigated by the steps component that makes it clear there is more to come.
* Forking the code for these flows seems like the right call as we want to A/B test this and do more changes later on. We just need to remove the two buttons on the top to start the new embedded flows before merging the feature branch into main for the initial release that will still link to the existing look
* Should a click outside of the modal close it? I expected it would, but no strong opinion
* I guess with the mentioned usability issues you are referring to having steps instead of a scrollable view, is that right? If yes, could you create an issue for this follow-up and add it to the board and the "Embed and improve logs" project?
* Could you add a quick note to the plugins readme to explain this component structure? It's probably not intuitive as we don't use this kind of backwards compatibility much and I don't want other devs to get confused when they need to touch it. While on it, could you also add a note on how to toggle the launchdarkly switch locally for testing?

flash1293 commented:
Would appreciate a quick look by @mykolaharmash or @justinkambic as well, but looks pretty good to me as a next step towards the planned look and feel 🎉

thomheymann commented:
> Thanks for this, @thomheymann
> 
> Some comments:
> 
> * Honestly even the current state doesn't look too bad to me - the scrollable content in the modal is somewhat mitigated by the steps component that makes it clear there is more to come.
> * I guess with the mentioned usability issues you are referring to having steps instead of a scrollable view, is that right? If yes, could you create an issue for this follow-up and add it to the board and the "Embed and improve logs" project?

@flash1293 My concern is that the existing flows haven't been designed to be shown inside a modal window. Generally speaking the contents of a modal should fit inside the window without scrolling so that important elements aren't hidden. In our case this issue is particular bad due to the way the forms are laid out. For example on the install agent step the only visible content on smaller screens are the two green success callouts telling the user that the "system integration has been installed" and that the "api key has been created" alongside a green "Explore logs" confirmation button. To someone unfamiliar with the flow that makes it seem like everything has been setup successfully and they might click "Explore logs" without even having started onboarding their data. 13" laptops are still common and not everyone runs their browser in full screen mode even on a desktop so we can't just design for the best case scenario. 

If we want to show these flows in a modal we should change the steps from vertical to horizontal and simplify them by removing any unnecessary steps and elements. The flows are currently designed to educate users and show too many technical details. I think this can be simplified further which would make it more suitable for modal windows and smaller screens as well. By embedded these flows as is however we are making the current situation worse so I would not feel comfortable releasing the new landing page without also working on the onboarding flows themselves. 

I'll work with design to address these.  

> * Should a click outside of the modal close it? I expected it would, but no strong opinion

I think in our case it is preferred for an outside click not to close the modal since users will be switching back and forth between terminal and browser window so it would be very easy to lose their progress accidentally. 

> * Forking the code for these flows seems like the right call as we want to A/B test this and do more changes later on. We just need to remove the two buttons on the top to start the new embedded flows before merging the feature branch into main for the initial release that will still link to the existing look
> * Could you add a quick note to the plugins readme to explain this component structure? It's probably not intuitive as we don't use this kind of backwards compatibility much and I don't want other devs to get confused when they need to touch it. While on it, could you also add a note on how to toggle the launchdarkly switch locally for testing?

Good idea. I'll add a README how to run the new landing page. I think it's too early days for dev docs though and I'm not even sure how many of these UI components we'll keep yet.

flash1293 commented:
Agreed with all of the above, one comment to 

> By embedded these flows as is however we are making the current situation worse so I would not feel comfortable releasing the new landing page without also working on the onboarding flows themselves

What I had in mind is to only release the first two steps (initial question and rendering of integration tiles), then navigate the user to the current full-screen version of the system logs and custom logs flow. The modal-embedded flows can still be in the code, as the user won't see them. I guess what we would need to change on this PR to make that work is to use a different path for the embedded vs non-embedded flows. Wdyt?

mykolaharmash commented:
Tested locally and functionally everything looks good ✨ Regarding the modal/no-modal discussion, I think it worth syncing with Christian as he showed me latest mockups for Nginx quick-start flow earlier today and there is no modal anymore and the whole thing looks very much like iteration over the existing UI, so it might make sense to incorporate what he came up with into existing flows already.

Commit:
Merge branch 'feat/obs-onboarding-flow' of https://github.com/elastic/kibana into 178948-embed-logs-flow

Commit:
Add instructions

Commit:
Fix i18n

thomheymann committed:
Merge branch 'feat/obs-onboarding-flow' into 178948-embed-logs-flow

flash1293 commented:
@thomheymann is this ready for another look?

thomheymann commented:
> @thomheymann is this ready for another look?

Yeh, unless we want to wait with this until the Grid has been merged

flash1293 commented:
Sorry, didn't get to review today's, will do first thing tomorrow morning

Commit:
Remove modal

thomheymann commented:
I've removed the modal and moved these flows onto a separate page as per latest designs. These are not final but it's a better starting point than the modals. 

<img width="1457" alt="Screenshot 2024-04-10 at 08 22 26" src="https://github.com/elastic/kibana/assets/190132/011aec81-d300-476e-98df-b7c342173f59">

flash1293 commented:
Thanks for this @thomheymann - the embedding into the new layout looks good to me. We talked about a bunch of improvements of the flow itself (simplifying the messaging etc.), but that will happen separately, right?

One thing I'm not sure about is the back button - this is currently not persisting the choice for the initial question which I guess we should make sure of. Options I can see here:
* Make the choice of the first question a navigation event (and a separate route) instead of keeping it as a local state in `OnboardingFlowForm`
* Move this state up above the routing layer so it stays in place

I don't have a strong opinion on how this is set up, the routing seems useful for the planned deep-linking features we talked about before, but it's nothing that can relatively easily be changed later on.

In general, let's try to split up the work we want to get merged for the first phase vs. what we are doing as subsequent iterations.

thomheymann commented:
> Thanks for this @thomheymann - the embedding into the new layout looks good to me. We talked about a bunch of improvements of the flow itself (simplifying the messaging etc.), but that will happen separately, right?

We are still working on the flows which will require a fair amount of work so I would not block this PR by the outcome of that. 

> One thing I'm not sure about is the back button - this is currently not persisting the choice for the initial question which I guess we should make sure of. Options I can see here:
> 
> * Make the choice of the first question a navigation event (and a separate route) instead of keeping it as a local state in `OnboardingFlowForm`
> * Move this state up above the routing layer so it stays in place
> 
> I don't have a strong opinion on how this is set up, the routing seems useful for the planned deep-linking features we talked about before, but it's nothing that can relatively easily be changed later on.
> 
> In general, let's try to split up the work we want to get merged for the first phase vs. what we are doing as subsequent iterations.

Good point, I'll go ahead and make that change

flash1293 commented:
> Good point, I'll go ahead and make that change

@justinkambic FYI - this is about introducing separate routes for the answer of the first question so the back button works right, as Thom is taking it on you can disregard my comment on your PR

Commit:
Move state to URL and expose experimental flow behind new path

Commit:
Merge branch 'feat/obs-onboarding-flow' of https://github.com/elastic/kibana into 178948-embed-logs-flow

mykolaharmash commented on x-pack/plugins/observability_solution/observability_onboarding/public/application/experimental_onboarding_flow.tsx:
```diff
@@ -29,6 +30,9 @@ import { CustomLogsPanel } from './quickstart_flows/custom_logs';
   const history = useHistory();
 
+  const [queryParams] = useSearchParams();
+  const queryString = queryParams.size ? `?${queryParams.toString()}` : '';
```
I believe `location.search` either from `useLocation()` or just `window.location` would produce the same thing here

mykolaharmash commented on x-pack/plugins/observability_solution/observability_onboarding/public/application/experimental_onboarding_flow.tsx:
```diff
@@ -100,14 +104,17 @@ export function ExperimentalOnboardingFlow() {
 const BackButton = () => {
-  const history = useHistory();
+  const navigate = useNavigate();
+  const [queryParams] = useSearchParams();
+  const queryString = queryParams.size ? `?${queryParams.toString()}` : '';
```
Same here about `location.search`

mykolaharmash commented on x-pack/plugins/observability_solution/observability_onboarding/public/plugin.ts:
```diff
@@ -120,6 +120,42 @@ export class ObservabilityOnboardingPlugin
         },
         visibleIn: [],
       });
+
```
Do we still need the logic above to get the `observability_onboarding.experimental_onboarding_flow_enabled'` config? This now should be handled by the new URL path, right?

flash1293 replied:
We still need to for A/B testing capabilities.

mykolaharmash reviewed (commented):
Thank you for creating the new path 🙏 I added just few minor comments.

mykolaharmash commented:
@thomheymann @flash1293 do you think we could switch the target branch to `main` here?

flash1293 reviewed (approved):
LGTM, let's try to get everything merged together quickly so we can develop against main.

Commit:
Added suggestion from code review

flash1293 commented:
CI failures look unrelated
